core/src/main/java/com/linecorp/armeria/client/AbstractDnsResolverBuilder.java

```diff
@@ -0,0 +1,558 @@
+    private long queryTimeoutMillisForEachAttempt = -1;
+
+    private boolean recursionDesired = true;
+    private int maxQueriesPerResolve = -1;
+    private int maxPayloadSize = 4096;
+    private boolean optResourceEnabled = true;
+
+    private HostsFileEntriesResolver hostsFileEntriesResolver = HostsFileEntriesResolver.DEFAULT;
+    private DnsServerAddressStreamProvider serverAddressStreamProvider =
+            DnsServerAddressStreamProviders.platformDefault();
+    @Nullable
+    private DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory;
+    private boolean dnsQueryMetricsEnabled = true;
+    private List<String> searchDomains = DnsUtil.defaultSearchDomains();
+    private int ndots = DnsUtil.defaultNdots();
+    private boolean decodeIdn = true;
+
+    @Nullable
+    private MeterRegistry meterRegistry;
+
+    /**
+     * Creates a new instance.
+     */
+    protected AbstractDnsResolverBuilder() {}
+
+    /**
+     * Sets if this resolver should generate detailed trace information in exception messages so that
+     * it is easier to understand the cause of resolution failure. This flag is enabled by default.
+     */
+    public AbstractDnsResolverBuilder traceEnabled(boolean traceEnabled) {
+        this.traceEnabled = traceEnabled;
+        return this;
+    }
+
+    /**
+     * Sets the timeout of the DNS query performed by this resolver.
+     * {@code 0} disables the timeout.
+     * If unspecified, {@value #DEFAULT_QUERY_TIMEOUT_MILLIS} ms will be used.
+     */
+    public AbstractDnsResolverBuilder queryTimeout(Duration queryTimeout) {
+        requireNonNull(queryTimeout, "queryTimeout");
+        checkArgument(!queryTimeout.isNegative(), "queryTimeout: %s (expected: >= 0)", queryTimeout);
+        return queryTimeoutMillis(queryTimeout.toMillis());
+    }
+
+    /**
+     * Returns the timeout of the DNS query performed by this resolver in milliseconds.
+     */
+    protected final long queryTimeoutMillis() {
+        return queryTimeoutMillis;
+    }
+
+    /**
+     * Sets the timeout of the DNS query performed by this resolver in milliseconds.
+     * {@code 0} disables the timeout.
+     * If unspecified, {@value #DEFAULT_QUERY_TIMEOUT_MILLIS} ms will be used.
+     */
+    public AbstractDnsResolverBuilder queryTimeoutMillis(long queryTimeoutMillis) {
+        checkArgument(queryTimeoutMillis >= 0, "queryTimeoutMillis: %s (expected: >= 0)", queryTimeoutMillis);
+        this.queryTimeoutMillis = queryTimeoutMillis;
+        return this;
+    }
+
+    /**
+     * Sets the timeout of each DNS query performed by this endpoint group.
+     * This option is useful if you want to set a timeout for each
+     * <a href="https://en.wikipedia.org/wiki/Search_domain">search domain</a> resolution.
+     * If unspecified, the value of {@link #queryTimeout(Duration)} is used.
+     */
+    public AbstractDnsResolverBuilder queryTimeoutForEachAttempt(Duration queryTimeoutForEachAttempt) {
```

Comment:
Q) Is there a need for both `queryTimeoutMillisForEachAttempt` and `queryTimeoutMillis`?
I first thought `queryTimeoutMillis` is a timeout for all queries, but functionally there seems to be no difference 🤔

Reply:
https://github.com/line/armeria/blob/f9f7ca2dde816bd405cf81cfcf388c31e8ce8639/core/src/main/java/com/linecorp/armeria/client/AbstractDnsResolverBuilder.java#L531-L539
`queryTimeoutForEachAttempt` will be passed to Netty DNS resolver if specified.